src/index.js

```diff
@@ -58,40 +31,63 @@ const moneyFormat = locale => ({
   }
 })
 
-const applyIntlToTranslations = (translations, locale, format) => (
-  transform(translations, (result, entry, key) => {
-    result[key] = new IntlFormat(entry, kebabCase(locale), { ...moneyFormat(locale), ...format })
-  })
-)
-
-const initializeTranslations = (translations, format) => (
-  transform(translations, (result, localeTranslations, locale) => {
-    result[locale] = applyIntlToTranslations(localeTranslations, locale, format)
-  })
-)
-
-const templateReducer = (values) => (result, v, k) => {
-  result[k] = v.format(values)
-  return result
+const createWithTranslation = (globalTranslations = {}, defaultLocale = DEFAULT_LOCALE) => {
+  const withTranslation = ({ translations, format = {} }) => {
+
+    return Component => class WrappedComponent extends React.Component {
+      static defaultProps = {
+        translations: {}
+      }
+
+      shouldComponentUpdate () {
+        return false
+      }
+
+      buildTranslations = locale => merge(
+        {},
+        globalTranslations[locale],
+        translations[defaultLocale],
+        translations[locale],
+        this.props.translations[locale]
+      )
+
+      getTranslateFunc = locale => {
+        const formats = {
+          ...moneyFormat(locale),
+          ...format
+        }
+
+        const allTranslations = this.buildTranslations(locale)
+
+        return function translate (key) {
+          const value = get(allTranslations, key)
+
+          if (isNumber(value) || isString(value)) {
+            const result = new IntlFormat(formats, value, locale)
+            return result.format(this.props)
+          }
+
+          return value
+        }
+      }
+
+      render () {
+        return (
+          <TranslationConsumer>
+            {locale => <Component translate={this.getTranslateFunc(locale)} />}
```

Comment:
If I get this correctly, the translations are no longer initialized before the wrapper function is created. I am not sure about the performance implications of (re-) initializing them at runtime on each props change.

Reply:
It was like that before; you need the locale, which only exists at runtime. But still, I don't like it either and initially tried to avoid recalculation by using shouldComponentUpdate. We could still do a shallow comparison on the translation props and only recalculate them, when they changed. That's pretty much what `_.memoize` does here. Only call `merge` when locale or the translation prop changed.